### ISSUE TITLE
Distinguishing stateful and stateless nodes

### ISSUE THREAD

**Issue by seefeldb** [google/labs-prototypes#48]
Right now they look the same, but I think only `localMemory` is actually stateful. For all the others the state is entirely captured by the incoming edges, and the same underlying handler works for all instances. For stateful nodes, instances have to be distinguished (is the suggestion in #37 related to this?).

One way we could make this explicit is by introducing an explicit `@self` wire that loops back to itself. The latest value would be passed in with each invocation. Internally, this could just be a pointer into another data structure, it doesn't have to be the actual full state.

This allows distinguishing two such nodes in a single graph. And we can keep one handler instance per node type.

It also opens the door for e.g. a `@clone` or `@mirror` endpoint that `@self` can be wired to. This can be useful when creating subagents that fork a memory, etc.

This makes all data flows explicit in the graph, which is important for the safety work.

One downside is that this self-wire has to be added when building a graph. Maybe that's actually good though, making the distinction with stateless nodes explicit. Or we add a helper that automatically adds a missing self-wire?

WDYT @dglazkov ?

**Comment by dglazkov** [comment]
This sounds very intriguing.

**Comment by dglazkov** [comment]
Just thinking out loud, not sure if there are any conclusions: In cases where the node calls out to some server API, the concept of "stateless" could shift to the server.

Example: I could send multiple 'text' values  to a vector store node, and each results in `text` value being added to the underlying vector store. It's stateless in itself, but there's definitely state being preserved in the vector store.

**Comment by seefeldb** [comment]
Yes, that's where I was actually hoping for a bit of symmetry with `@self`: Such a node would have an input that specifies the address of the server API, auth tokens, etc., which essentially identifies the instance. So both are then stateless, and both tie into longer living state via one of their inputs.

FWIW: For the safety work, there will have to be a pass that reveals these implicit data flows. Implicit, because we could have data flowing from a store node to a retrieval node without there being a path between them in the graph. Understanding that identity giving input will be important, ranging from a simple understanding that two stores are indeed the same to much more sophisticated parsing of the actual queries and figuring out which nodes write things that which other other nodes read.

The constraint solver would then figure out what safety properties a writer needs to have, depending on readers elsewhere. (Rough analogy: Type inference that span all the way from reader to (implicitly) storage layer to writer)


FWIW, this could look like
- a single node representing the server address, wired to both writers and readers
  - that single node taking some namespace-y thing as input, thus creating a unique space
- a writer node having an output that is just the address, that then goes to the reader, roughly mimicing the `@mirror` idea implied above (what's nice here is that the flow is then in graph again)

**Comment by dglazkov** [comment]
I still don't understand the idea behind `@self`. My mind just keeps slipping off it.

**Comment by seefeldb** [comment]
Hmm, oh, I think we don't need to special case `@self`, an optional input is probably sufficient. Consider for example rewriting `local-memory` as `append` that takes in an optional `list` and required `values` and outputs `list`, and then use it by wiring the output `list` right back to the input `list`:
```mermaid
flowchart LR
   A(new data) -->|values| B[append]
   B -. list .-> B
   B -->|list| C(whole list consumer)
```

Then all nodes' state is fully defined by their inputs, no internal state. This means no lifecycle management for nodes. One handler per node type, not node. Which is what the code currently supports.

Unless I misread the code, right now all `local-memory` nodes share the same `context` defined in the closure when the kit gets loaded. So if I write into one `local-memory` node, the values get added to all of the nodes in the graph. That's surprising. To change that we could either do the above (move the state out) or flag this node type as stateful and use a factory, remember, that, etc. -- maybe there are other simpler ways?

Now let's see how a vector database could work. For in memory, it could be:
```mermaid
flowchart TD
    subgraph indexing loop
    D(new data) -->|documents| B
    end
    A[create-vector-db] --o|db| B[add-to-vector-db]
    A --o|db| C[query-vector-db]
    subgraph querying loop
    E(query) --> C
    C -->|result| F(results)
    end
 ```

For a permanent one, `create-vector-db` would consume an address of the database.

And then `add-to-vector` could also output a convenience `db` output, which would be identical to the input, but sent on every output, in case the query is fixed and should be rerun on every database change:
```mermaid
flowchart TD
    subgraph indexing loop
    D(new data) -->|documents| B
    end
    A[create-vector-db] --o|db| B[add-to-vector-db]
    B --o|db| C[query-vector-db]
    subgraph static query
    E(query) --> C
    C -->|result| F(results)
    end
```

**Comment by seefeldb** [comment]
This has implications for #29.

The questions are:
 1) If graphs are nodes, could we package up composite nodes and have some syntactic sugar around `include`-ing them?
 2) Are `include`s stateful or not?

E.g. it would be neat to package up the pattern in the previous comment as a `local-memory` graph and then add some syntactic sugar so that we can refer to a `local-memory` node that is actually an `include` of that graph.

But: The new idea there is that the list is passed to itself. If `include` recreates the graph on each invocation, then there would only be an empty list every time. So, two options:
 1) We use the self-pass pattern at the graph level
 2) We have some sort of stateful `include`

I think (2) brings a lot of complexity to the engine, and it reintroduces the ambiguity about stateful nodes and also breaks the nice graphs-are-nodes property. Hence I'm leaning towards (1). (Note also that if the state was hidden in a node, that there would be no way for state to survive multiple `include` passes unless we add even more lifecycle management!)

But: What if there is a more complex graph with multiple such self-pass stateful nodes? We'd have to self-pass all that state around via the `input` and `output` nodes. And more problematically: That would expose the inner workings, breaking the abstraction.

Instead we'd probably want little helper nodes that collect all the states and sends them to output and do the reverse on input. But now we'll have to also wire every self-pass state to to these helper nodes. That's too much work and too fiddly:

We need some syntactic sugar to make this easy for developers/graph authors. I propose that this is the general pattern:
 1) Keep the underlying engine as simple as possible, including trying to keep any lifecycle management out
 2) Use the pattern above to create state
 3) Add syntactic sugar to make the most common cases easy.

AI: Write up a proposal for that syntactic sugar. @dglazkov WDYT?

**Comment by dglazkov** [comment]
Let's lean on the idea that there aren't actually any stateful nodes.

I am going to start with converting `local-memory` to `appender`.
